File: AlgoExpert/Dynamic Programming/Medium/Min Number Of Coins For Change.java
import java.util.*;

class Program {
  public static int minNumberOfCoinsForChange(int n, int[] denoms) {
    // Write your code here.
		int[] numberOfCoins = new int[n + 1];
		Arrays.fill(numberOfCoins, Integer.MAX_VALUE);
		numberOfCoins[0] = 0;
		
		for(int denom: denoms){
			for(int amount = 1; amount < numberOfCoins.length; amount++){
				if(denom <= amount){
					int newNum;
					if(numberOfCoins[amount - denom] == Integer.MAX_VALUE)
						newNum = numberOfCoins[amount - denom];
					else
						newNum = numberOfCoins[amount - denom] + 1;
					if(newNum < numberOfCoins[amount]){
						numberOfCoins[amount] = newNum;
					}
				}
			}
		}
		
    return numberOfCoins[n] != Integer.MAX_VALUE ? numberOfCoins[n]: -1;
  }
}

File: Python/AlgoExpert/Dynamic Programming/Medium/minNumberOfCoinsForChange.py
def minNumberOfCoinsForChange(n, denoms):
    # Write your code here.
    num_of_coins = [float("inf") for i in range(n + 1)]
    num_of_coins[0] = 0

    for amount in range(n + 1):
        for denom in denoms:
            if denom > amount:
                continue
            if amount - denom >= 0:
                num_of_coins[amount] = min(num_of_coins[amount], num_of_coins[amount - denom] + 1)

    return num_of_coins[n] if num_of_coins[n] != float("inf") else -1
